File: apps/desktop/src-tauri/src/commands/rdp.rs
use tauri::{AppHandle, Manager, State};

use crate::rdp::{RdpBounds, RdpManager, RdpRuntimeStatus};

use super::run_blocking;

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn create_rdp(
    app: AppHandle,
    state: State<'_, RdpManager>,
    session_id: String,
    host: String,
    port: u16,
    username: String,
    admin_session: bool,
    bounds: RdpBounds,
) -> Result<(), String> {
    let manager = state.inner().clone();
    let cancellation = manager.begin(&session_id)?;
    let cleanup_manager = manager.clone();
    let cleanup_session_id = session_id.clone();
    let cleanup_cancellation = cancellation.clone();
    let result = on_main_thread(app, "RDP 创建", move |app| {
        #[cfg(windows)]
        {
            let window = app.get_webview_window("main").ok_or("找不到主窗口")?;
            let parent = window
                .hwnd()
                .map_err(|error| format!("无法读取主窗口：{error}"))?;
            let scale = window
                .scale_factor()
                .map_err(|error| format!("无法读取窗口缩放：{error}"))?;
            manager.create(
                parent,
                session_id,
                &host,
                port,
                &username,
                admin_session,
                bounds.physical(scale),
                cancellation,
            )
        }
        #[cfg(not(windows))]
        {
            let _ = (
                app,
                session_id,
                host,
                port,
                username,
                admin_session,
                bounds,
                cancellation,
            );
            manager.unsupported()
        }
    })
    .await;
    if result.is_err() {
        cleanup_manager.cancel_connecting(&cleanup_session_id, &cleanup_cancellation);
    }
    result
}

#[tauri::command]
pub async fn resize_rdp(
    app: AppHandle,
    state: State<'_, RdpManager>,
    session_id: String,
    bounds: RdpBounds,
    visible: bool,
) -> Result<(), String> {
    let manager = state.inner().clone();
    on_main_thread(app, "RDP 调整", move |app| {
        #[cfg(windows)]
        {
            let window = app.get_webview_window("main").ok_or("找不到主窗口")?;
            let scale = window
                .scale_factor()
                .map_err(|error| format!("无法读取窗口缩放：{error}"))?;
            manager.resize(&session_id, bounds.physical(scale), visible)
        }
        #[cfg(not(windows))]
        {
            let _ = (app, session_id, bounds, visible);
            manager.unsupported()
        }
    })
    .await
}

#[tauri::command]
pub async fn get_rdp_status(
    app: AppHandle,
    state: State<'_, RdpManager>,
    session_id: String,
) -> Result<RdpRuntimeStatus, String> {
    let manager = state.inner().clone();
    on_main_thread(app, "RDP 状态读取", move |_app| {
        #[cfg(windows)]
        {
            manager.status(&session_id)
        }
        #[cfg(not(windows))]
        {
            let _ = session_id;
            manager.unsupported()?;
            unreachable!()
        }
    })
    .await
}

#[tauri::command]
pub async fn close_rdp(
    app: AppHandle,
    state: State<'_, RdpManager>,
    session_id: String,
) -> Result<(), String> {
    let manager = state.inner().clone();
    on_main_thread(app, "RDP 关闭", move |_app| {
        #[cfg(windows)]
        {
            manager.close(&session_id)
        }
        #[cfg(not(windows))]
        {
            let _ = session_id;
            manager.unsupported()
        }
    })
    .await
}

async fn on_main_thread<T, F>(
    app: AppHandle,
    label: &'static str,
    operation: F,
) -> Result<T, String>
where
    T: Send + 'static,
    F: FnOnce(AppHandle) -> Result<T, String> + Send + 'static,
{
    run_blocking(label, move || {
        let (sender, receiver) = std::sync::mpsc::sync_channel(1);
        let task_app = app.clone();
        app.run_on_main_thread(move || {
            let _ = sender.send(operation(task_app));
        })
        .map_err(|error| format!("无法调度{label}：{error}"))?;
        receiver
            .recv()
            .map_err(|_| format!("{label}没有返回结果"))?
    })
    .await
}

File: apps/desktop/src/ipc/rdp.ts
import { invoke } from "@tauri-apps/api/core";
import { invalidResponse, normalizeIpcError } from "./errors";

export interface RdpBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface RdpCreateRequest {
  sessionId: string;
  host: string;
  port: number;
  username: string;
  adminSession: boolean;
  bounds: RdpBounds;
}

export type RdpRuntimeState = "initializing" | "connecting" | "connected" | "disconnected";

export interface RdpRuntimeStatus {
  state: RdpRuntimeState;
  disconnectReason: string | null;
  focused: boolean;
}

const COMMANDS = {
  create: "create_rdp",
  resize: "resize_rdp",
  status: "get_rdp_status",
  focus: "focus_rdp",
  close: "close_rdp",
} as const;

export const rdpApi = {
  create(request: RdpCreateRequest) {
    return invokeCommand<void>(COMMANDS.create, { ...request });
  },

  resize(sessionId: string, bounds: RdpBounds, visible: boolean) {
    return invokeCommand<void>(COMMANDS.resize, { sessionId, bounds, visible });
  },

  status(sessionId: string) {
    return invokeCommand<unknown>(COMMANDS.status, { sessionId }).then(decodeRuntimeStatus);
  },

  focus(sessionId: string) {
    return invokeCommand<void>(COMMANDS.focus, { sessionId });
  },

  close(sessionId: string) {
    return invokeCommand<void>(COMMANDS.close, { sessionId });
  },
};

async function invokeCommand<T>(command: string, args: Record<string, unknown>) {
  try {
    return await invoke<T>(command, args);
  } catch (error) {
    throw normalizeIpcError(error);
  }
}

function decodeRuntimeStatus(value: unknown): RdpRuntimeStatus {
  if (!isRecord(value)
    || !isRuntimeState(value.state)
    || (value.disconnectReason !== null && typeof value.disconnectReason !== "string")
    || typeof value.focused !== "boolean") {
    throw invalidResponse("RDP 状态格式无效");
  }
  return {
    state: value.state,
    disconnectReason: value.disconnectReason,
    focused: value.focused,
  };
}

function isRuntimeState(value: unknown): value is RdpRuntimeState {
  return value === "initializing"
    || value === "connecting"
    || value === "connected"
    || value === "disconnected";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
